### PR TITLE
[DTRA] Maryia/DTRA-970/build: packages audit, upgrading to node 18 & webpack 5

### PR DESCRIPTION
**Pull request by maryia-deriv**
Changes:

#### Includes:

1. Upgrading to `webpack 5`,
2. Upgrading to `node 18`,
3. Replacing `node-sass` with modern `sass`,
4. Replacing `react-svg-loader` with modern `@svgr/webpack`,
5. Updating other packages versions in order to reduce dependency issues coming from outdated versions,
6. Fixing a code scanning issue [Prototype-polluting assignment](https://github.com/deriv-com/smarttrader/security/code-scanning/62)

## Type of change

-   [ ] Bug fix
-   [ ] New feature
-   [ ] Update feature
-   [x] Refactor code
-   [ ] Translation to code
-   [ ] Translation to crowdin
-   [ ] Script configuration
-   [ ] Improve performance
-   [ ] Style only
-   [x] Dependency update
-   [x] Documentation update
-   [ ] Release

### CODE REVIEW

File: .github/workflows/release_staging.yml [hunk 1]
```diff
@@ -9,11 +9,11 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - name: Checkout
-        uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11
+        uses: actions/checkout@v4
```
Comment (balakrishna-deriv) [inline .github/workflows/release_staging.yml]: we should be using the sha as it is the security best practice. So, can we revert these changes? :)

**Review by akmal-deriv — APPROVED** [review]

File: src/javascript/config.js [hunk 2]
```diff
@@ -55,9 +55,6 @@ const getAppId = () => {
     } else if (/staging-smarttrader\.deriv\.app/i.test(window.location.hostname)) { // TODO: [app-link-refactor] - Remove backwards compatibility for `deriv.app`
         window.localStorage.removeItem('config.default_app_id');
         app_id = 22169;
-    } else if (/staging-smarttrader\.deriv\.com/i.test(window.location.hostname)) {
-        window.localStorage.removeItem('config.default_app_id');
-        app_id = 22169;
```
Comment (maryia-deriv) [inline src/javascript/config.js:60]: removing since it's a duplicate

File: build/webpack/plugins.js [hunk 3]
```diff
@@ -29,25 +52,7 @@ const getPlugins = (app, grunt) => ([
             }),
         ]
         : [
-            function() {
-                this.plugin('watch-run', (watching, callback) => {
```
Comment (maryia-deriv) [inline build/webpack/plugins.js:33]: replaced this part with a custom `WatchRunPlugin` since `this.plugin` no longer exists in webpack.

File: .stylelintrc.js [hunk 4]
```diff
@@ -2,85 +2,33 @@ module.exports = {
...
-        'string-quotes'                                    : 'single',
         'time-min-milliseconds'                            : 100,
-        'unit-case'                                        : 'lower',
-        'unit-whitelist'                                   : ['px', 'em', 'rem', '%', 'vw', 'vh', 'deg', 'ms', 's', 'fr'],
+        'unit-allowed-list'                                : ['px', 'em', 'rem', '%', 'vw', 'vh', 'deg', 'ms', 's', 'fr'],
         'value-keyword-case'                               : 'lower',
-        'value-list-comma-newline-after'                   : 'never-multi-line',
-        'value-list-comma-newline-before'                  : 'never-multi-line',
-        'value-list-comma-space-after'                     : 'always',
-        'value-list-comma-space-before'                    : 'never',
```
Comment (maryia-deriv) [inline .stylelintrc.js:84]: Removed all the rules that are deprecated in stylelint v15.
Please let me know if any concerns.

File: .eslintrc.js [hunk 5]
```diff
@@ -12,35 +12,38 @@ module.exports = {
...
+        'no-trailing-spaces'            : ['error', { skipBlankLines: true }],
+        'object-curly-spacing'          : ['error', 'always', { arraysInObjects: true, objectsInObjects: true }],
+        'one-var'                       : ['error', { initialized: 'never', uninitialized: 'always' }],
+        'prefer-destructuring'          : 0,
+        'space-in-parens'               : ['error', 'never'],
+        'space-infix-ops'               : 'error',
+        'space-unary-ops'               : 'error',
+        'no-multiple-empty-lines'       : ['error', { 'max': 1, 'maxEOF': 1 }],
+        'prefer-exponentiation-operator': 0,
+        'prefer-object-spread'          : 0,
```
Comment (maryia-deriv) [inline .eslintrc.js:39]: Turned off `prefer-exponentiation-operator` & `prefer-object-spread` since this code base is legacy and it only supports ES2015 (E6) syntax.
Please let me know if any concerns.

**Review by kate-deriv — APPROVED** [review]

File: src/sass/_common/external/gridle/_mixins.scss [hunk 6]
```diff
@@ -290,14 +290,16 @@
 	@if gridle_has_state($stateName) {
 		// extend the "at-root" version for the .state-{stateName} classes
 		@if $states-classes and str-index(unquote("#{&}"), unquote('.state-#{$_gridle_current_stateName}')) {
-			// @at-root (without: media) { // seems not needed here, cause issues with libsass
+			@at-root (without: media) {
```
Comment (ako-deriv) [inline src/sass/_common/external/gridle/_mixins.scss:293]: doesnt this have impact on the styles?

File: src/sass/_common/common.scss [hunk 7]
```diff
@@ -12,7 +12,7 @@ body {
     background: repeat-x bottom;
     overflow-x: hidden;
     font-family: $FONT-STACK !important;
-    text-rendering: optimizeLegibility;
+    text-rendering: optimizelegibility;
```
Comment (ako-deriv) [inline src/sass/_common/common.scss]: as far as checked 
```suggestion
    text-rendering: optimizeLegibility;
```
this one is the correct one.

File: src/javascript/app/pages/trade/common.js [hunk 8]
```diff
@@ -367,9 +367,7 @@ const commonTrading = (() => {
     };
 
     const reloadPage = () => {
-        Defaults.remove('market', 'underlying', 'formname',
-            'date_start', 'time_start', 'expiry_type', 'expiry_date', 'expirt_time', 'duration_units', 'diration_value',
-            'amount', 'amount_type', 'currency', 'prediction');
+        Defaults.remove('market', 'underlying', 'formname', 'date_start', 'time_start', 'expiry_type', 'expiry_date', 'expirt_time', 'duration_units', 'diration_value', 'amount', 'amount_type', 'currency', 'prediction');
```
Comment (ako-deriv) [inline src/javascript/app/pages/trade/common.js]:
```suggestion
        Defaults.remove('market', 'underlying', 'formname', 'date_start', 'time_start', 'expiry_type', 'expiry_date', 'expiry_time', 'duration_units', 'diration_value', 'amount', 'amount_type', 'currency', 'prediction');
```
I know that this is something that you didn't change, but its obviously a typo, is it possible for you to check it?

File: scripts/extract_js_texts.js [hunk 9]
```diff
@@ -34,7 +33,10 @@ const source_strings = {};
 const ignored_list   = {};
 const invalid_list   = {};
 
-let this_app_name;
+let this_app_name, emphasize;
+import('emphasize').then(module => {
```
Comment (ako-deriv) [inline scripts/extract_js_texts.js:37]: why do we need to import it like this?

File: scripts/extract_js_texts.js [hunk 10]
```diff
@@ -3,7 +3,6 @@
 /* eslint-disable no-console */
 const BabelParser = require('@babel/parser');
 const color       = require('cli-color');
-const emphasize   = require('emphasize');
 const estraverse  = require('estraverse');
 const fs          = require('fs');
 const Path        = require('path');
```
Comment (ako-deriv) [inline scripts/extract_js_texts.js:8]: if we have import here maybe we can change all these requires here to import

File: package.json [hunk 11]
```diff
@@ -20,97 +20,90 @@
...
-    "webpack-dev-server": "3.1.14",
-    "webpack-manifest-plugin": "2.0.4",
+    "webpack": "5.81.0",
+    "webpack-bundle-analyzer": "4.3.0",
     "webpack-merge": "4.1.4",
-    "ws": "7.0.0"
+    "ws": "8.16.0"
   },
   "dependencies": {
+    "@babel/polyfill": "7.12.1",
```
Comment (ako-deriv) [inline package.json:102]: this needs to move to dev deps

File: build/webpack/config_common.js [hunk 12]
```diff
@@ -1,28 +1,29 @@
...
     optimization: {
-        namedChunks: true,
-        minimize   : true,
-        minimizer  : [
-            new UglifyJsPlugin({
-                test     : /\.min\.js/,
-                exclude  : /vendors~/,
-                parallel : true,
-                sourceMap: true,
+        chunkIds : 'named',
```
Comment (ako-deriv) [inline build/webpack/config_common.js:17]: image
just as a suggestion we can also use natural for this value as well.
still depends on the requirements that you have in the team and your priorities.

File: build/config/constants.js [hunk 13]
```diff
@@ -39,7 +39,6 @@ const release_config = {
 
 const node_modules_paths = {
     binary_style: 'node_modules/@binary-com/binary-style',
-    deriv_p2p   : 'node_modules/@deriv/p2p',
```
Comment (ako-deriv) [inline build/config/constants.js:42]: why this needs to removed?

File: .stylelintrc.js (hunk 4 above)
Comment (ako-deriv) [reply .stylelintrc.js:84]: doesnt it break the style lints?
maybe any major changes?

**Review by ali-hosseini-deriv — CHANGES_REQUESTED** [review]

File: build/config/constants.js (hunk 13 above)
Comment (maryia-deriv) [reply build/config/constants.js:42]: `deriv_p2p` was causing dependency issues and it is no longer used in the repo)

File: build/webpack/config_common.js (hunk 12 above)
Comment (maryia-deriv) [reply build/webpack/config_common.js:17]: @ali-hosseini-deriv I chose 'named' as a replacement to the unsupported property that I had to remove - `namedChunks: true` - in order to keep existing functionality :)

File: package.json (hunk 11 above)
Comment (maryia-deriv) [reply package.json:102]: According to the docs, it should be installed in the regular 'dependencies':
Screenshot 2024-03-21 at 12 14 16 PM

I replaced `babel-polyfill` with it. `babel-polyfill` was also located inside the regular 'dependencies'.

File: scripts/extract_js_texts.js (hunk 10 above)
Comment (maryia-deriv) [reply scripts/extract_js_texts.js:8]: The following error doesn't let me replace 'require's with 'import's in this module:
>"Cannot use import declarations in modules that export using CommonJS (module.exports = 'foo' or exports.bar = 'hi')eslint(import/no-import-module-exports)"

File: scripts/extract_js_texts.js (hunk 9 above)
Comment (maryia-deriv) [reply scripts/extract_js_texts.js:37]: I had to dynamically import only `emphasize` because only for this package I got the following error:
>"Error [ERR_REQUIRE_ESM]: require() of ES Module /Users/maryiahryshchuk/Projects/smarttrader-binary-static/node_modules/emphasize/index.js from /Users/maryiahryshchuk/Projects/smarttrader-binary-static/scripts/extract_js_texts.js not supported.
Instead change the require of index.js in /Users/maryiahryshchuk/Projects/smarttrader-binary-static/scripts/extract_js_texts.js to a dynamic import() which is available in all CommonJS modules."

An attempt to use 'import's and 'export default' causes another error:
>"Cannot use import statement outside a module"

because CommonJS is used in this project, not ESM.

File: src/sass/_common/common.scss (hunk 7 above)
Comment (maryia-deriv) [reply src/sass/_common/common.scss]: unfortunately, with stylelint version 15, the correct one is `optimizelegibility` :)
>"Running "stylelint:all" (stylelint) task
>
>  src/sass/_common/common.scss:15:21
>  ✖  15:21  Expected optimizeLegibility to be optimizelegibility"

But it seems that `stylelint` is wrong in this case...
Removing the existing rule `'value-keyword-case' : 'lower'` helps to enforce `optimizeLegibility`.
 I'll review all the stylelint rules in the project again. Thank you)

File: src/javascript/app/pages/trade/common.js (hunk 8 above)
Comment (maryia-deriv) [reply src/javascript/app/pages/trade/common.js]: I found another typo besides this one. Therefore, I've introduce a constant for this strings :)

File: scripts/extract_js_texts.js (hunk 10 above)
Comment (farhan-nurzi-deriv) [reply scripts/extract_js_texts.js:8]: You can change the last 3 lines to these exports, then it will works with import i guess

```
// export parse function in function declaration
// export function getTexts(app_name = this_app_name) {};
// export function getErrorsCount(app_name = this_app_name) {};
```

**Review by balakrishna-deriv — APPROVED** [review]

File: scripts/extract_js_texts.js (hunk 10 above)
Comment (maryia-deriv) [reply scripts/extract_js_texts.js:8]: this doesn't work unfortunately. Also, as i mentioned [here](https://github.com/deriv-com/smarttrader/pull/617#discussion_r1533480060) :

An attempt to use 'import's and 'export default' causes another error:

>"Cannot use import statement outside a module"

because CommonJS is used in this project, not ESM.

File: src/sass/_common/external/gridle/_mixins.scss (hunk 6 above)
Comment (maryia-deriv) [reply src/sass/_common/external/gridle/_mixins.scss:293]: @ali-hosseini-deriv I uncommented this because of the error:
>"You may not @extend selectors across media queries."

`@at-root (without: media)` helps us move this `@extend` outside of the `@media`, right where it's supposed to be. Styles look good after this change. I wouldn't use this approach if the classname was regular, not made of sass-variables. :)

**Review by ali-hosseini-deriv — APPROVED** [review]

File: .stylelintrc.js (hunk 4 above)
Comment (maryia-deriv) [reply .stylelintrc.js:84]: there is no automated linting like prettifier on this project, and after removing these rules it looks good, nothing is broken, no errors. Unfortunately, there are no equivalents to these deprecated rules in stylelint 15.
They really look useless.
Should I leave them and silence deprecation warning instead?

**Review by farhan-nurzi-deriv — APPROVED** [review]
